README.md
![SDK](https://shields.microej.com/endpoint?url=https://repository.microej.com/packages/badges/sdk_5.7.json)
![ARCH](https://shields.microej.com/endpoint?url=https://repository.microej.com/packages/badges/arch_7.18.json)

# Overview

MicroEJ Core Engine Abstraction Layer implementation for uC-OS3.

This module implements the `LLMJVM` Low Level API for MicroEJ VEE Ports connected to a Board Support Package based on [uC-OS3](https://github.com/weston-embedded/uC-OS3/).

See the MicroEJ documentation for a description of the `LLMJVM` functions:
- [LLMJVM: MicroEJ Core Engine](https://docs.microej.com/en/latest/PlatformDeveloperGuide/appendix/llapi.html#llmjvm-microej-core-engine)
- [MicroEJ Core Engine: Implementation](https://docs.microej.com/en/latest/PlatformDeveloperGuide/coreEngine.html#implementation)

# Usage

1. Install ``src`` and ``inc`` directories in your Board Support Package. They can be automatically downloaded using the following command lines:
   ```sh
    svn export --force https://github.com/MicroEJ/AbstractionLayer-Core-uC-OS3/trunk/src/main/c/inc [path_to_bsp_directory]    
    svn export --force https://github.com/MicroEJ/AbstractionLayer-Core-uC-OS3/trunk/src/main/c/src [path_to_bsp_directory]
   ```

2. Implement the MicroEJ time functions, as described in [microej_time.h](./src/main/c/inc/microej_time.h).

3. The `LLMJVM_IMPL_scheduleRequest` function in [LLMJVM_uCOS3.c](./src/main/c/src/LLMJVM_uCOS3.c) uses a software timer. In order to correctly schedule MicroEJ threads, check the following elements in the uC-OS3 configuration file:

   - Timers are enabled with `#define OS_CFG_TMR_EN 1`
   - `OS_CFG_TMR_TASK_PRIO` is smaller than MicroEJ task priority
   - `OS_CFG_TMR_TASK_RATE_HZ`: can depend on the application, if it needs a 1 ms precision then the tick rate would be 1000 Hz, the recommended value is between 100 Hz and 1000 Hz

# Requirements

None.

# Validation

This Abstraction Layer implementation can be validated in the target Board Support Package using the [MicroEJ Core Validation](https://github.com/MicroEJ/VEEPortQualificationTools/tree/master/tests/core/java-testsuite-runner-core) VEE LLMJVM_UCOS3_CONFIGURATION Qualification Tools project.

Here is a non exhaustive list of tested environments:
- Hardware:
  - STMicroelectronics STM32F746-DISCO
- Compilers / Integrated Development Environments:
  - IAR Embedded Workbench 9.30.1
- uC-OS3 version:
  - uC-OS3 3.08.02

# MISRA Compliance

This Abstraction Layer implementation is MISRA-compliant (MISRA C 2012) with the following observed deviations:
| Deviation | Category |                                                 Justification                                                 |
|:---------:|:--------:|:-------------------------------------------------------------------------------------------------------------:|
| Rule 8.4  | Required | ``LLMJVM_IMPL`` functions are declared in ``LLMJVM_impl.h``. 														   | 
| Rule 21.6 | Required | ``printf`` from ``<stdio.h>`` is used by ``LLMJVM_ASSERT_TRACE_OUTPUT``. The output function can be configured in [LLMJVM_uCOS3_configuration.h](./src/main/c/inc/LLMJVM_uCOS3_configuration.h).	| 
| Rule 8.9  | Advisory | ``LLMJVM_UCOS3_next_wake_up_time`` is needed as a global variable.              |
| Rule 2.7  | Advisory | The uC-OS3 timer callback function need to comply with the ``OS_TMR_CALLBACK_PTR`` definition.              |
| Rule 8.7  | Advisory | ``LLMJVM_IMPL_getCurrentTime`` needs to be accessed by the Core Engine.            |  

# Dependencies

- MicroEJ Architecture ``7.x`` or higher.
- FreeRTOS ``3.08.02`` or higher.

# Source

N/A.

# Restrictions

None.

---

_Copyright 2023 MicroEJ Corp. All rights reserved._
_Use of this source code is governed by a BSD-style license that can be found with this software._
 


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2023-05-31

### Added

- Initial revision.

---
_Copyright 2023 MicroEJ Corp. All rights reserved._
_Use of this source code is governed by a BSD-style license that can be found with this software._

src/main/c/inc/microej_time.h
/*
 * C
 *
 * Copyright 2018-2023 MicroEJ Corp. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be found with this software.
 */

/**
 * @brief This is the definition of the MicroEJ time API.
 * 
 * To implement this API the use of a hardware timer is recommended. The validation is done with the Core Engine Test Suite: https://github.com/MicroEJ/VEEPortQualificationTools/tree/master/tests/core/
 * 
 * There are 2 types of time in MicroEJ:
 * - The platform time: an arbitrary time that is only relevant to measure elapsed time
 * - The application time: the time elapsed since midnight, January 1, 1970 UTC, this time should be set by the MicroEJ application
 * 
 * 
 */

#ifndef MICROEJ_TIME_H
#define MICROEJ_TIME_H

#include <stdint.h>

#ifdef __cplusplus
    extern "C" {
#endif

/**
 * @brief Initialisation of the timebase used by MicroEJ. Initialize in this function everything that is needed to implement the API.
 * 
 */
void microej_time_init(void);

/**
 * @brief Gets the current platform or application time.
 * 
 * @param is_platform_time set to 1 to get the platform time or set to 0 to get the time elapsed since midnight, January 1, 1970 UTC.
 * @return int64_t the time in milliseconds.
 */
int64_t microej_time_get_current_time(uint8_t is_platform_time);

/**
 * @brief Gets an arbitrary time.
 * 
 * @return int64_t the time in nanoseconds. The time should have a nanosecond precision but not necessarily nanosecond resolution.
 */
int64_t microej_time_get_time_nanos(void);

/**
 * @brief Sets the application time.
 * 
 * @param time value in milliseconds of the time elapsed since midnight, January 1, 1970 UTC.
 */
void microej_time_set_application_time(int64_t time);

/**
 * @brief Converts a time in milliseconds to a number of system ticks.
 * 
 * @param time the time to convert in milliseconds,
 * @return int64_t the number of ticks.
 */
int64_t microej_time_time_to_tick(int64_t time);

#ifdef __cplusplus
    }
#endif

#endif /* _MICROEJ_TIME_H */


src/main/c/inc/LLMJVM_uCOS3_configuration.h
/*
 * C
 *
 * Copyright 2023 MicroEJ Corp. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be found with this software.
 */
#ifndef LLMJVM_UCOS3_CONFIGURATION
#define LLMJVM_UCOS3_CONFIGURATION

#ifdef __cplusplus
    extern "C" {
#endif

/**
 * @brief Compatibility sanity check value.
 * This define value is checked in the implementation to validate that the version of this configuration
 * is compatible with the implementation.
 *
 * This value must not be changed by the user of the CCO.
 * This value must be incremented by the implementor of the CCO when a configuration define is added, deleted or modified.
 */
#define LLMJVM_UCOS3_CONFIGURATION_VERSION (1)

#define LLMJVM_TRACE_ENABLED

#ifndef LLMJVM_TRACE_ENABLED
#define LLMJVM_ASSERT_TRACE_OUTPUT(...) ((void)0U)
#else
// cppcheck-suppress [misra-c2012-21.6] printf is used by default by LLMJVM_ASSERT_TRACE_OUTPUT, it can be changed below.
#include <stdio.h>
#define LLMJVM_ASSERT_TRACE_OUTPUT                \
  (void) printf("[LLMJVM ASSERT FAILED] ");       \
  (void) printf
#endif

#ifdef __cplusplus
    }
#endif

#endif /* LLMJVM_UCOS3_CONFIGURATION_H */

src/main/c/src/LLMJVM_uCOS3.c
/*
 * C
 *
 * Copyright 2014-2023 MicroEJ Corp. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be found with this software.
 */

/**
 * @file
 * @brief LLMJVM implementation over UC-OS III.
 * @author MicroEJ Developer Team
 * @version 1.0.0
 */


/*
 *********************************************************************************************************
 *                                             INCLUDE FILES
 *********************************************************************************************************
 */

#include <stdlib.h>
#include <stdbool.h>

// Micrium µCosIII includes
#include "os_cfg_app.h"
#include "os.h"
#include "lib_def.h"

#include "LLMJVM_impl.h"
#include "microej.h"
#include "microej_time.h"
#include "LLMJVM_uCOS3_configuration.h"

#if !defined LLMJVM_UCOS3_CONFIGURATION_VERSION
     #error "Undefined LLMJVM_UCOS3_CONFIGURATION_VERSION, it must be defined in microvg_configuration.h"
#endif

#if defined LLMJVM_UCOS3_CONFIGURATION_VERSION && LLMJVM_UCOS3_CONFIGURATION_VERSION != 1
     #error "Version of the configuration file LLMJVM_UCOS3_CONFIGURATION.h is not compatible with this implementation."
#endif

#ifdef __cplusplus
    extern "C" {
#endif

/*
 *********************************************************************************************************
 *                                       LOCAL GLOBAL VARIABLES
 *********************************************************************************************************
 */

/* Defines -------------------------------------------------------------------*/

/* Initial delay of the timer */
#define LLMJVM_UCOS3_WAKEUP_TIMER_DELAY_MS          ((int64_t)100)

/* The 'period' being repeated for the timer. */
#define LLMJVM_UCOS3_WAKEUP_TIMER_PERIOD_MS         ((int64_t)0)

/* Assert macro */
#define LLMJVM_ASSERT(x)                                          \
  if (!(x))                                                       \
  {                                                               \
      LLMJVM_ASSERT_TRACE_OUTPUT("%s, %d\n", __FILE__, __LINE__); \
      while (1){                                                  \
                                                                  \
      }                                                           \
                                                                  \
  }                                                               \
  else                                                            \
  {                                                               \
    ;                                                             \
  }                                                               \

/*
 * Shared variables
 */

/* Set to true when the timer expires, set to true when the timer is started. */
static volatile bool LLMJVM_UCOS3_timer_expired = false;

//Absolute time in ticks at which timer will be launched
// cppcheck-suppress [misra-c2012-8.9] this variable needs to be declared globally.
static int64_t LLMJVM_UCOS3_next_wake_up_time = INT64_MAX;

// Timer for scheduling next alarm
static OS_TMR LLMJVM_UCOS3_wake_up_timer = NULL;

/* Binary semaphore to wakeup MicroJVM */
static OS_SEM LLMJVM_UCOS3_Semaphore = NULL;

/* Private functions ---------------------------------------------------------*/

static void wake_up_timer_callback(void *p_tmr, void *p_arg);

/* Since LLMJVM_schedule() prototype does not match the timer callback prototype,
   we create a wrapper around it and check the ID of the timer. */
// cppcheck-suppress [misra-c2012-2.7, constParameter] Need to comply with OS_TMR_CALLBACK_PTR definition
static void wake_up_timer_callback(void *p_tmr, void *p_arg) {
    if (p_tmr == &LLMJVM_UCOS3_wake_up_timer) {
        LLMJVM_UCOS3_timer_expired = true;
        LLMJVM_schedule();
    }
}

/* Public functions ----------------------------------------------------------*/

/*
 * Implementation of functions from LLMJVM_impl.h
 * and other helping functions.
 */

// Creates the timer used to callback the LLMJVM_schedule() function.
// After its creation, the timer is idle.
int32_t LLMJVM_IMPL_initialize(void)
{  

  int32_t rslt = LLMJVM_OK;
  OS_ERR err;

  if (OSCfg_TmrTaskRate_Hz <= (OS_RATE_HZ) 0) {
      /* MicriumOS timer task disabled or not configured correctly. */
      rslt = LLMJVM_ERROR;
  }

  if (LLMJVM_OK == rslt) {
    /* Create a timer to scheduler alarm for the VM. Delay and period values are dummy initialization values which will never be used. */
    OSTmrCreate(&LLMJVM_UCOS3_wake_up_timer,
                "MicroJVM wake up",
                (OS_TICK)microej_time_time_to_tick(LLMJVM_UCOS3_WAKEUP_TIMER_DELAY_MS),
                (OS_TICK)microej_time_time_to_tick(LLMJVM_UCOS3_WAKEUP_TIMER_PERIOD_MS),
                OS_OPT_TMR_ONE_SHOT,
                wake_up_timer_callback,
                NULL,
                &err);

    if (OS_ERR_NONE != err) {
        rslt = LLMJVM_ERROR;
    }
  }

  if (LLMJVM_OK == rslt) {
    OSSemCreate(&LLMJVM_UCOS3_Semaphore,
                "MicroJVM wake up",
                (CPU_INT32U)0,
                &err);
    
    microej_time_init();
    
    if (OS_ERR_NONE != err) {
        rslt = LLMJVM_ERROR;
    }
  }

  return rslt;

}

// Once the task is started, save a handle to it.
int32_t LLMJVM_IMPL_vmTaskStarted(void)
{
	return LLMJVM_OK;
}

// Schedules requests from the VM
int32_t LLMJVM_IMPL_scheduleRequest(int64_t absoluteTime)
{
        int32_t rslt = LLMJVM_OK;
        OS_ERR err_set = OS_ERR_NONE;
        OS_ERR err_start = OS_ERR_NONE;
        OS_ERR err_stop = OS_ERR_NONE;
	      int64_t relativeTime;
        int64_t relativeTick;
        CPU_BOOLEAN xTimerStartResult;

  int64_t currentTime = LLMJVM_IMPL_getCurrentTime((uint8_t)MICROEJ_TRUE);

  relativeTime = absoluteTime - currentTime;
  /* Determine relative time/tick */
  relativeTick = microej_time_time_to_tick(relativeTime);

  if (relativeTick <= 0) {
    /* 'absoluteTime' has been reached yet */

    /* No pending request anymore */
    LLMJVM_UCOS3_next_wake_up_time = INT64_MAX;

    /* Stop current timer (no delay) */
    if (OS_TMR_STATE_RUNNING == LLMJVM_UCOS3_wake_up_timer.State) {
        OSTmrStop(&LLMJVM_UCOS3_wake_up_timer, OS_OPT_TMR_NONE, NULL, &err_stop);
        LLMJVM_ASSERT(OS_ERR_NONE == err_stop);
    }

    /* Notify the VM now */
    rslt = LLMJVM_schedule();
  } else if ((true == LLMJVM_UCOS3_timer_expired) ||
        (absoluteTime < LLMJVM_UCOS3_next_wake_up_time) ||
        (LLMJVM_UCOS3_next_wake_up_time <= currentTime)) {
      /* We want to schedule a request in the future but before the existing request
         or the existing request is already done */

      /* Save new alarm absolute time */
      LLMJVM_UCOS3_next_wake_up_time = absoluteTime;

      /* Stop current timer (no delay) */
      if (OS_TMR_STATE_RUNNING == LLMJVM_UCOS3_wake_up_timer.State) {
          OSTmrStop(&LLMJVM_UCOS3_wake_up_timer, OS_OPT_TMR_NONE, NULL, &err_stop);
          LLMJVM_ASSERT(OS_ERR_NONE == err_stop);
      }
      LLMJVM_UCOS3_timer_expired = false;

      /* Schedule the new alarm */
      OSTmrSet(&LLMJVM_UCOS3_wake_up_timer,
               (OS_TICK)relativeTick,
               (OS_TICK)0,
               wake_up_timer_callback,
               NULL,
               &err_set);
      xTimerStartResult = OSTmrStart(&LLMJVM_UCOS3_wake_up_timer, &err_start);

      if ((DEF_TRUE != xTimerStartResult) ||
          (OS_ERR_NONE != err_set) ||
          (OS_ERR_NONE != err_start)) {
          rslt = LLMJVM_ERROR;
      }
  } else {
    /* else: there is a pending request that will occur before the new one -> do nothing */
  }
  
  return rslt;
}


// Suspends the VM task if the pending flag is not set
int32_t LLMJVM_IMPL_idleVM(void)
{
    OS_ERR err = OS_ERR_NONE;
    OSSemPend(&LLMJVM_UCOS3_Semaphore,
              (OS_TICK)0,
              OS_OPT_PEND_BLOCKING,
              NULL,
              &err);

    return (OS_ERR_NONE == err) ? (int32_t)LLMJVM_OK : (int32_t)LLMJVM_ERROR;

}

// Wakes up the VM task and reset next wake up time
int32_t LLMJVM_IMPL_wakeupVM(void)
{
      OS_ERR err = OS_ERR_NONE;
      OSSemPost(&LLMJVM_UCOS3_Semaphore,
          OS_OPT_POST_ALL,
          &err);

      return (OS_ERR_NONE == err) ? (int32_t)LLMJVM_OK : (int32_t)LLMJVM_ERROR;

}

// Clear the pending wake up flag
int32_t LLMJVM_IMPL_ackWakeup(void)
{
	return LLMJVM_OK;
}

int32_t LLMJVM_IMPL_getCurrentTaskID(void)
{
	return (int32_t)OSTCBCurPtr;
}

// Sets application time
void LLMJVM_IMPL_setApplicationTime(int64_t t)
{
    microej_time_set_application_time(t);
}

// Gets the system or the application time in milliseconds
// cppcheck-suppress [misra-c2012-8.7] this function needs to be accessed by the Core Engine, see LLMJVM_impl.h.
int64_t LLMJVM_IMPL_getCurrentTime(uint8_t sys)
{      
    return microej_time_get_current_time(sys);
}

// Gets the current system time in nanoseconds
int64_t LLMJVM_IMPL_getTimeNanos(void){
    return microej_time_get_time_nanos();
}

int32_t LLMJVM_IMPL_shutdown(void)
{
	// nothing to do
	return LLMJVM_OK;
}

#ifdef __cplusplus
    }
#endif